src/test/java/jenkins/plugins/logstash/LogstashWriterTest.java

```diff
@@ -0,0 +1,299 @@
+    // Verify results
+    // No error output
+    assertEquals("Results don't match", "", errorBuffer.toString());
+    verify(mockBuild).getLog(3);
+
+    verify(mockDao).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), Matchers.anyListOf(String.class));
+    verify(mockDao).push("{\"data\":{},\"message\":[\"test\"],\"source\":\"jenkins\",\"source_host\":\"http://my-jenkins-url\",\"@version\":1}");
+  }
+
+  @Test
+  public void writeSuccessConnectionBroken() throws Exception {
+    Mockito.doNothing().doThrow(new IOException("BOOM!")).doNothing().when(mockDao).push(anyString());
+    LogstashWriter los = createLogstashWriter(mockBuild, errorBuffer, "http://my-jenkins-url", mockDao, mockBuildData);
+
+
+    String msg = "test";
+    String exMessage = "[logstash-plugin]: Failed to send log data to REDIS:localhost:8080.\n" +
+      "[logstash-plugin]: No Further logs will be sent to localhost:8080.\n" +
+      "java.io.IOException: BOOM!\n";
+
+    errorBuffer.reset();
+
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertEquals("Results don't match", "", errorBuffer.toString());
+
+    // Break the dao connnection
+    errorBuffer.reset();
+
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertTrue("Results don't match", errorBuffer.toString().startsWith(exMessage));
+    assertTrue("Connection not broken", los.isConnectionBroken());
+
+    // Verify logs still write but on further calls are made to dao
+    errorBuffer.reset();
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertEquals("Results don't match", "", errorBuffer.toString());
+
+    //Verify calls were made to the dao logging twice, not three times.
+    verify(mockDao, times(2)).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), Matchers.anyListOf(String.class));
+    verify(mockDao, times(2)).push("{\"data\":{},\"message\":[\"test\"],\"source\":\"jenkins\",\"source_host\":\"http://my-jenkins-url\",\"@version\":1}");
+    verify(mockDao).getIndexerType();
+    verify(mockDao, times(2)).getDescription();
+  }
+
+  @Test
+  public void writeBuildLogGetLogError() throws Exception {
+    // Initialize mocks
+    when(mockBuild.getLog(3)).thenThrow(new IOException("Unable to read log file"));
+
+    LogstashWriter writer = createLogstashWriter(mockBuild, errorBuffer, "http://my-jenkins-url", mockDao, mockBuildData);
+    assertEquals("Errors were written", "", errorBuffer.toString());
+
+    // Unit under test
+    writer.writeBuildLog(3);
+
+    // Verify results
+    verify(mockBuild).getLog(3);
+
+    String errorMsg = "[logstash-plugin]: Unable to serialize log data.\n" +
+      "java.io.IOException: Unable to read log file\n";
+    assertThat("Wrong error message", errorBuffer.toString(), containsString(errorMsg));
+    verify(mockDao).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), anyListOf(String.class));
```

Comment:
Yeah, I can think of two options:
a) Use `ArgumentCaptor<List<String>>` for the logLines and search the list for the entry you expect
b) Create a new Matcher that extends `ArgumentMatcher<List<String>>` that does the same kind of search as in option a

Since this is only for verification at the end of the test, not test setup, using an ArgumentCaptor is  IMO better than an ArgumentMatcher. 

Add this import:
```
import static org.hamcrest.core.IsCollectionContaining.hasItem;
```

Add this member to the LogstashWriterTest:
```
// This goes with the mocks and gets initialized automagically by MockitoJUnitRunner
@Captor ArgumentCaptor<List<String>> logLinesCaptor;
```

Then line 296 becomes:
```
verify(mockDao).buildPayload(eq(mockBuildData), eq("http://my-jenkins-url"), logLinesCaptor.capture());
assertThat("The exception was not sent to Logstash", logLinesCaptor.getValue(), hasItem(errorMsg.split("\n")[0]));
```